### PR TITLE
Fix: wrong indent at tagged template in indent (fixes #12122)

### LINKED ISSUE #12122: indent issue with tagged template literals
**Tell us about your environment**

* **ESLint Version:** 6.2.0
* **Node Version:** 10.16.3
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "es6": true,
    "jest/globals": true
  },
  "plugins": ["jest"],
  "rules": {
    "indent": [
      "error", 2, {
      "SwitchCase": 1,
      "VariableDeclarator": 1,
      "outerIIFEBody": 1,
      "FunctionDeclaration": {
        "parameters": 1,
        "body": 1
      },
      "FunctionExpression": {
        "parameters": 1,
        "body": 1
      },
      "CallExpression": {
        "arguments": 1
      },
      "ArrayExpression": 1,
      "ObjectExpression": 1,
      "ImportDeclaration": 1,
      "flatTernaryExpressions": false,
      "ignoredNodes": [
        "JSXElement", "JSXElement > *", "JSXAttribute", "JSXIdentifier", "JSXNamespacedName", "JSXMemberExpression", "JSXSpreadAttribute", "JSXExpressionContainer", "JSXOpeningElement", "JSXClosingElement", "JSXText", "JSXEmptyExpression", "JSXSpreadChild"
      ],
      "ignoreComments": false
    }]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js

describe('Indentation Bug', () => {

  test.each`
    a    | b
    ${1} | ${1}
  `('$a == $b', ({ a, b }) => {
    expect(a).toBe(b); // error  Expected indentation of 2 spaces but found 4  indent
  }); // error  Expected indentation of 0 spaces but found 2  indent
});
```

```bash
eslint .
```

**What did you expect to happen?**
I would expect the code indentation to be valid.

**What actually happened? Please include the actual, raw output from ESLint.**
Following eslint errors are displayed:

```
8:1  error  Expected indentation of 2 spaces but found 4  indent
9:1  error  Expected indentation of 0 spaces but found 2  indent
```

I've created a small repo where you can easily reproduce the issue.
https://github.com/msvab/eslint-tagged-literal-indent-bug

**Are you willing to submit a pull request to fix this bug?**
I can try, but I have zero knowledge on eslint internals.

### PR DESCRIPTION
**Pull request by yeonjuan**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

File: lib/rules/indent.js [hunk 1]
```diff
@@ -922,7 +922,15 @@ module.exports = {
 
             parameterParens.add(openingParen);
             parameterParens.add(closingParen);
-            offsets.setDesiredOffset(openingParen, sourceCode.getTokenBefore(openingParen), 0);
+
+            let offsetFrom = sourceCode.getTokenBefore(openingParen);
+
+            // If the node's callee is a TaggedTemplateExpression, use callee's tag as a offset from.
```
Comment (kaicataldo) [inline lib/rules/indent.js]: I think we can get rid of this comment - the code actually reads a lot like English in this case 😊

File: lib/rules/indent.js [hunk 2]
```diff
@@ -922,7 +922,15 @@ module.exports = {
 
             parameterParens.add(openingParen);
             parameterParens.add(closingParen);
-            offsets.setDesiredOffset(openingParen, sourceCode.getTokenBefore(openingParen), 0);
+
+            let offsetFrom = sourceCode.getTokenBefore(openingParen);
```
Comment (kaicataldo) [inline lib/rules/indent.js]: Can we name this `offsetToken` or `offsetFromToken`?

**Comment by mysticatea** [comment]
I'm not sure... but I couldn't reproduce it on v6.7.2: [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgaW5kZW50OiBbZXJyb3IsIDRdICovXG5cbm9mZnNldFplcm8oKCkgPT4ge1xuICAgIHRhZ2BvZmZzZXQgaXMgXG4tLS0tYWx3YXlzXG4tLS0temVyb2AoKCkgPT4ge1xuICAgICAgICBmb28oKVxuICAgICAgICAvLyBvZmZzZXQgemVyb1xuICAgIH0pO1xufSk7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

**Comment by mysticatea** [comment]
I could reproduce it if an interpolation exists: [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgaW5kZW50OiBbZXJyb3IsIDRdICovXG5cbm9mZnNldFplcm8oKCkgPT4ge1xuICAgIHRhZ2BvZmZzZXQgaXMgXG4tLS0tYWx3YXlzXG4tLS0tJHtudWxsfVxuLS0tLXplcm9gKCgpID0+IHtcbiAgICAgICAgZm9vKClcbiAgICAgICAgLy8gb2Zmc2V0IHplcm9cbiAgICB9KTtcbn0pOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

Therefore, I think that this bug exists around the logic of template literals rather than function calls. I'm suspecting that the indent rule doesn't set desired offsets to the tokens of `TemplateElement` nodes.

**Comment by yeonjuan** [comment]
@mysticatea 
> I'm suspecting that the indent rule doesn't set desired offsets to the tokens of TemplateElement nodes.
oh yes, It's right. I think I should change the code in this pr. thanks!

**Comment by yeonjuan** [comment]
@mysticatea 
I changed the test cases, including expressions within template literals.
And I tried to set the desired offset to the token but I couldn't because it makes another change in the bellow cases.

```js
const a = 2,
    template = `
    multiline ${
    expression} template literal`; // Error; 
```

so, what about just add logic within the function which checking function call(same as current pr)?

**Comment by mysticatea** [comment]
I got it. In that case, I'm OK with fixing the logic of function calls.

**Comment by yeonjuan** [comment]
plz, don't merge it. I'm sorry. It needs more check.

**Comment by yeonjuan** [comment]
@kaicataldo 
Sorry for changing the code after review. But I realized that the previous version does not work correctly. The previous version uses `tag` as an offset token. It could solve the bug in a tagged template indent.

```js

tag`
   ${a}
  `(() => {
      tag`
        ${a}
        `(() => {

    });
});

```

but it had problems with member expression with tagged template literal.

```js
tag
   .member`
   ${a}
  `(() => {

}); // expected indent, but no error.
```

So, I changed the code to use a token before `quasi` as an offset token.  So now, it solves the usual  `tagged template` indent issues. You can check it in test cases.

https://github.com/eslint/eslint/blob/c780efba577ea663d79961d15edcf6ee70b5da75/tests/lib/rules/indent.js#L5418-L5438

https://github.com/eslint/eslint/blob/c780efba577ea663d79961d15edcf6ee70b5da75/tests/lib/rules/indent.js#L5518-L5524

I re-request a review. Sorry for wasting your efforts.

File: lib/rules/indent.js [hunk 3]
```diff
@@ -922,7 +922,11 @@ module.exports = {
 
             parameterParens.add(openingParen);
             parameterParens.add(closingParen);
-            offsets.setDesiredOffset(openingParen, sourceCode.getTokenBefore(openingParen), 0);
+
+            const offsetAfterToken = node.callee.type === "TaggedTemplateExpression" ? node.callee.quasi : openingParen;
```
Comment (kaicataldo) [inline lib/rules/indent.js]: It looks like `offsets.setDesiredOffset()` is expecting a token. For consistency, should we do the following?

```suggestion
            const offsetAfterToken = node.callee.type === "TaggedTemplateExpression" ? sourceCode.getFirstToken(node.callee.quasi) : openingParen;
```

Comment (yeonjuan) [reply lib/rules/indent.js]: @kaicataldo yes! thanks for review

**Pull request merged**